infrastructure/infrastructure/src/main/java/alabno/wserver/StudentJob.java

```diff
@@ -29,4 +36,73 @@ public String readPostProcessorOutput() {
     public String getJsonLocation() {
         return jsonLocation;
     }
+
+    public void amend(String fileName, int lineno, String annType, String annotation) {
+        if ("ok".equals(annotation)) {
+            annotation = "";
+        }
+        
+        String desiredFile = toAbsolute(fileName);
+        
+        // Read the postpro file
+        String postproContent = readPostProcessorOutput();
+        
+        JsonParser parser = new JsonParser(postproContent);
+        
+        // get the annotations part
+        JsonArrayParser annotations = parser.getArrayParser("annotations");
+
+        for (JsonParser ann : annotations) {
+            
+            String aFileName = ann.getString("filename");
+            int aLineNo = ann.getInt("lineno");
+            
+            if (desiredFile.equals(aFileName) && lineno == aLineNo) {
+                JSONObject amended = ann.getObject();
+                amended.put("errortype", annType);
+                amended.put("text", annotation);
+                
+                rewriteJson(parser);
+                return;
+            }
+        }
+        
+        // If loop reaches, nothing was found
+        System.out.println("Could not find entry to be amended. Adding a new one...");
+        JSONObject newAnn = new JSONObject();
+        newAnn.put("errortype", annType);
+        newAnn.put("filename", desiredFile);
+        newAnn.put("lineno", lineno);
+        newAnn.put("charno", 1);
+        newAnn.put("text", annotation);
+        
+        // append to the existing array
+        JSONArray annotationsArray = parser.getArray("annotations");
+        if (annotationsArray == null) {
+            annotationsArray = new JSONArray();
+        }
+        annotationsArray.add(newAnn);
+        parser.getObject().put("annotations", annotationsArray);
+        
+        rewriteJson(parser);
+        return;
+        
+    }
+    
+    private void rewriteJson(JsonParser parser) {
+        try {
+            SubprocessUtils.call("rm " + jsonLocation);
+            File jsonFile = new File(jsonLocation);
+            PrintWriter writer = new PrintWriter(jsonFile);
+            writer.print(parser.getObject().toJSONString());
+            writer.close();
+        } catch (Exception e) {
+            e.printStackTrace();
+        }
+    }
+
+    String toAbsolute(String relpath) {
+        // utility to convert to absolute path based on the relative one
+        return jsonLocation.substring(0, Math.max(0, jsonLocation.length() - 17)) + "/" + relpath;
```

Comment:
removes the part of the path containing output.json, to get the directory name only